src/hit/api/params.cpp

```diff
@@ -3,10 +3,105 @@
+    // the total bit size of the _product_ of each "prime", so there are no security implications if
+    // we accidentally generate a composite.
+    int miller_rabin_iters = 25;
+
+    // true if x \in mods, false otherwise
+    bool contains(uint64_t x, const vector<Modulus> &mods) {
+        for (const auto &m : mods) {
+            if (x == m.value()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // return the representative of x mod n in (-n/2, n/2]
+    uint64_t repr(uint64_t x, uint64_t n) {
+        uint64_t cmod = x % n;
+        return (cmod <= (n / 2)) ? cmod : (cmod - n);
+    }
+
+    // return the closest integer to x such that x = 1 mod n
+    uint64_t to1Coset(uint64_t x, uint64_t n) {
+        // note that n is always even in our uses case.
+        // if x <= kn+n/2 (for some k), then repr(x,n) > 0,
+        //   so we return kn+1, which is closer than (k+1)n+1.
+        // if x > kn+n/2 (for some k), then repr(x,n) < 0,
+        //   so we return (k+1)n+1, which is at least as close as kn+1.
+        return x - repr(x, n) + 1;
+    }
+
+    // Find the smallest prime larger than `x` that is not in `mods`.
+    uint64_t nextPrime(uint64_t x, uint64_t n, const vector<Modulus> &mods) {
+        // find a number near x that is congruent to 1 mod n
+        uint64_t next_test_val = to1Coset(x, n);
+
+        // to1Coset can return a value smaller than x
+        if (next_test_val < x) {
+            next_test_val += n;
+        }
+
+        while (!miller_rabin_test(next_test_val, 25) || contains(next_test_val, mods)) {
+            next_test_val += n;
+        }
+        return next_test_val;
+    }
+
+    // Find the largest prime smaller than `x` that is not in `mods`.
+    uint64_t prevPrime(uint64_t x, uint64_t n, const vector<Modulus> &mods) {
+        // find a number near x that is congruent to 1 mod n
+        uint64_t next_test_val = to1Coset(x, n);
+
+        // to1Coset can return a value larger than x
+        if (next_test_val > x) {
+            next_test_val -= n;
+        }
+
+        while (!miller_rabin_test(next_test_val, 25) || contains(next_test_val, mods)) {
+            next_test_val -= n;
+        }
+        return next_test_val;
+    }
+
+    // implements the prime selection algorithm in https://eprint.iacr.org/2020/1118, Algorithm 3
+    vector<Modulus> reducedErrorPrimes(int poly_mod_degree, const vector<int> &modulus_vec) {
+        int num_moduli = modulus_vec.size();
+        vector<Modulus> primes(num_moduli);
+        // m = 2 * poly_mod_degree is the cyclotomic index of the ring
+        // all primes should be congruent to 1 mod m so that the cyclotomic polynomial splits completely mod q
+        uint64_t m = 2 * poly_mod_degree;
+
+        // generate a prime for keyswitching
+        // this is independent of the ciphertext chain
+        primes[num_moduli - 1] = prevPrime(to1Coset(pow(2, modulus_vec[num_moduli - 1]), m), m, primes);
```

Comment:
It looks like we call `to1Coset` both before passing in the first parameter and after receiving it. Is this duplicate work?